{% extends "web_home.html" %}

{% load staticfiles %}
{% load add_class %}
{% load add_class %}
{% load search_result_formatter %}

{% block content %}

<div class="mane mane-intro container" style="padding-top:20px;">

    <h1>MANE Collaboration</h1>

<div class="intro">
  <p>The <span class="bold_font">M</span>atched <span class="bold_font">A</span>nnotation from <span class="bold_font">N</span>CBI and <span class="bold_font">E</span>MBL-EBI (<span class="bold_font">MANE</span>) project
 is a collaboration that aims to converge on human gene annotation and to produce a genome wide transcript set that includes pairs of RefSeq (NM) and Ensembl/GENCODE (ENST) transcripts that are 100% identical.</p>
</div>

<div class="row" style="margin-bottom: 30px">


       <div class="col-md-4 col-sm-4">
     <a href="/web/access_mane_data/"><button type="button" class="btn btn-default btn-circle btn-xl center-block"><i class="glyphicon glyphicon-file"></i></button>
     <h3 class="text-center">Access MANE sequences</h3>
     <p>Download or view the MANE transcripts<br>(e.g. bulk download, REST API, browser etc.)</p>
         </a>
  </div>

   <div class="col-md-4 col-sm-4">
    <a href="/web/about_mane_collaboration/"><button type="button" class="btn btn-default btn-circle btn-xl center-block"><i class="glyphicon glyphicon-info-sign"></i></button>
 <h3 class="text-center">About the MANE Collaboration</h3>
  <p>MANE Projects and background</p>
       </a>
  </div>

   <div class="col-md-4 col-sm-4">
     <a href="/web/mane_feedback/"><button type="button" class="btn btn-default btn-circle btn-xl center-block"><i class="glyphicon glyphicon-envelope"></i></button>
 <h3 class="text-center">Feedback</h3>
  <p>We welcome input and comments</p>
       </a>
    </div>


</div>

<div class="row">
    <div class="col-md-12">


        <div class="heading-pt">
            <div class="glyph-container"><span class="glyphicon glyphicon-record"></span></div>
            <div class="text-container">MANE project outputs</div>
        </div>

                <h4>MANE Select</h4>
        <ul>
            <li>One single transcript at each protein-coding locus across the genome that is representative of biology at that locus</li>
            <li>Useful to provide a consistent default transcript per locus across resources</li>
        </ul>

        <h4>MANE Plus Clinical</h4>
        <ul>
            <li>Additional transcripts per locus necessary to support clinical variant reporting, e.g. which contain known pathogenic variants</li>
        </ul>

                <div class="heading-pt">
            <div class="glyph-container"><span class="glyphicon glyphicon-record"></span></div>
            <div class="text-container">What is unique about the MANE project?</div>
        </div>

        <h4>Highest quality annotation</h4>
        <ul>
            <li>Created jointly by <a class="icon-external-link" href="https://www.gencodegenes.org/about.html" target="_blank">Ensembl/GENCODE</a> at <a class="icon-external-link" href="https://www.ebi.ac.uk/" target="_blank">EMBL-EBI</a>, and <a class="icon-external-link" href="https://www.ncbi.nlm.nih.gov/refseq/about/" target="_blank">RefSeq</a> at <a class="icon-external-link" href="https://www.ncbi.nlm.nih.gov/" target="_blank">NCBI</a></li>
            <li>Identified using independent computational methods complemented by manual review of key loci and discussion by expert curator scientists</li>
        </ul>

        <h4>Identical transcript sequences from 5' to 3'</h4>
        <ul>
            <li>For each representative transcript, the pair of sequences from RefSeq and Ensembl/GENCODE are identical (5’ UTR, coding region and 3’ UTR) </li>
            <li>Transcript sequences retain their original identifiers / stable IDs e.g. <a href="https://www.ncbi.nlm.nih.gov/nuccore/NM_014491.4" target="_blank">NM_014491.4</a>; <a href="https://www.ensembl.org/Homo_sapiens/Transcript/Summary?g=ENSG00000128573;r=7:114414815-114693765;t=ENST00000350908" target="_blank">ENST00000350908.9</a></li>
          <li>All transcript sequences perfectly match the GRCh38 reference assembly</li>
        </ul>

        <h4>Connect the past, present and future of clinical variant reporting</h4>
        <ul>
            <li>Part of the effort to rationalise minor differences in RefSeq and Ensembl/GENCODE gene sets</li>
          <li>Facilitates unambiguous data exchange between specific RefSeq and Ensembl/GENCODE transcripts and the GRCh38/hg38 reference genome assembly</li>
          <li>Evolved to improve on previous consensus CDS (<a class="icon-external-link" href="https://www.ncbi.nlm.nih.gov/projects/CCDS">CCDS)</a> project and <a class="icon-external-link" href="http://www.lrg-sequence.org/">LRG project</a> (for legacy and community-requested stable sequences for reporting variants)</li>
          <li>Ensures compatibility with genome reference assembly-based NGS variant reporting systems</li>
        </ul>

        <h4>Well-supported</h4>
        <ul>
           <li>Annotated in the <a class="icon-external-link" href="https://www.ensembl.org/" target="_blank">Ensembl</a>, <a class="icon-external-link" href="https://www.ncbi.nlm.nih.gov/" target="_blank">NCBI</a> and <a class="icon-external-link" href="https://genome.ucsc.edu/" target="_blank">UCSC</a> genome browsers to allow visualisation in genomic context, with all other existing annotations</li>
         <li>Included in other resources e.g. ClinVar and gnomAD
          <li>Compatible with the Human Genome Variation Society (HGVS) nomenclature and supported by nomenclature checker systems (e.g. <a class="icon-external-link" href="https://www.mutalyzer.nl/" target="_blank">Mutalyzer</a>, <a class="icon-external-link" href="https://variantvalidator.org/" target="_blank">VariantValidator</a>)</li>
          <li>Generated and maintained by the <a class="icon-external-link" href="https://www.ncbi.nlm.nih.gov/" target="_blank">NCBI</a> and <a class="icon-external-link" href="https://www.ebi.ac.uk/" target="_blank">EMBL-EBI</a></li>
        </ul>



    </div>

</div>


 </div>

{% endblock %}