{% extends "web_home.html" %}

{% load staticfiles %}
{% load add_class %}
{% load add_class %}
{% load search_result_formatter %}

{% block content %}

<div class="container" style="padding-top:20px;">

<div class="row">

    <div class="col-md-12">

<h1>About the MANE Collaboration</h1>

        <p>The MANE project - The Matched Annotation from the National Center for Biotechnology Information (NCBI) and the European Molecular Biology Laboratories - European Bioinformatics Institute (EMBL-EBI) (MANE) is a collaborative project that aims to converge on human gene annotation to produce a genome wide transcript set that:</p>
            <ul>
                <li>
                    matches GRCh38
                </li>
                <li>
                    includes pairs of RefSeq (NM) and Ensembl/GENCODE (ENST) transcripts that are 100% identical (5’ UTR, coding region and 3’ UTR)
                </li>
                <li>
                    is high-confidence based on conservation, expression and overall LRG creation process – Locus Reference Genomic – LRG sequences provide a stable genomic DNA framework for reporting variants with a permanent ID and core content that never changes.
                </li>
            </ul>
        <p>Each transcript will retain its original identifier e.g. "ENST00000350908.9", or "NM_000088.3".</p>

        <h2>Background</h2>
        <p>Clinical data are more often reported using RefSeq (NM) transcripts, whereas large-scale genomics projects more often default to Ensembl’s ENSTs. This project will eliminate differences between RefSeq and Ensembl annotation for the MANE subset of transcripts to encourage and facilitate bi-directional exchange of data across RefSeq and Ensembl transcripts. Finally, the perfect alignment of all MANE transcripts to GRCh38 makes the set compatible with NGS-based sequencing technologies and other resources which default to GRCh38.</p>
    <p>The MANE Select set will serve as a default set across genomics resources. Thus our effort to select one high-quality transcript at all protein-coding loci, and to have this be consistent across all genomics resources, will give a consistent starting view of biology for researchers, whether the intent is to use it for reporting variants, comparative genomics or any other endeavour. For the MANE Plus Clinical, we have identified additional transcripts as a standard critical for variant reporting.</p>




        <h2> Current and future projects </h2>


        <h4>MANE Select&nbsp;<span class="label label-success">Current project</span></h4>
        <ul>
          <li>This set consists of one high-quality transcript at each protein-coding locus across the genome that is representative of the biology at that gene locus and is well-supported by experimental evidence.</li>
          <li> MANE Select transcripts are identified using computational methods complemented by manual review and discussion. The methods consider evidence of functional potential such as expression levels, evolutionary conservation and clinical significance.</li>
          <li>This set is useful as:</li>
            <ul>
         <li> a default for display on browsers,</li>
         <li>a standard for clinical reporting,</li>
         <li>a starting point for comparative or evolutionary genomics.</li>
            </ul>
        </ul>

    <div class="alert alert-warning" role="alert">
        <span class="glyphicon glyphicon-info-sign"></span>
          Note: All transcripts for a given locus should always be considered and we are certainly NOT saying that
      biology can be simplified to a single transcript.
    </div>



        <h4>MANE Plus Clinical&nbsp;<span class="label label-success">Current Project</span></h4>

        <ul>
          <li>Additional transcripts per locus necessary to support clinical variant reporting, for example transcripts containing known “Pathogenic” or “Likely Pathogenic” clinical variants not reportable using the MANE Select set. </li>
          <li>Note there may be additional clinically relevant transcripts in the wider  <a href="https://www.ncbi.nlm.nih.gov/refseq/about/" class="icon-external-link" target="_blank">RefSeq</a> and Ensembl/GENCODE sets but not yet in MANE. </li>
           <li> Collaboration with, and feedback from, the clinical community is an important aspect of the MANE project. Transcripts of clinical interest not included in our MANE sets will be considered upon request and in consultation with expert clinical scientists. Contact email <a href="mailto:mane-help@ebi.ac.uk">mane-help@ebi.ac.uk</a>.
    </li>
        </ul>


    <h4>MANE Clinical&nbsp;<span class="label label-info">Future project</span></h4>
        <ul>
          <li>Will aim to contain loci and transcripts that are relevant for clinical testing, including all the transcripts in MANE Plus Clinical. </li>
          <li>This project is on hold until there is more certainty as to which genes and transcripts are implicated in human disease, which will be done in consultation with clinical experts. </li>
        </ul>


    <h4>MANE Plus&nbsp;<span class="label label-info">Future project</span></h4>
        <ul>
          <li>An expansion of MANE transcripts beyond MANE Select and those required for clinical reporting (MANE Plus Clinical) to include transcripts that have additional well-supported features not found in other MANE transcript sets.</li>
          <li>The MANE Plus project is on hold while EMBL-EBI and NCBI work to improve their independent pipelines. We will:
            <ul>
              <li>explore and agree on thresholds for more automated methods for a transcript to be included</li>
              <li>mature our workflows to integrate long transcriptomic data from PacBio and Oxford Nanopore Technologies (ONT)</li>
              <li>update pipelines to utilise new data arising from the rapid technical development in the wider transcriptomic field (e.g. 5’ capping for PacBio and ONT, tagged full length ONT, direct RNA sequencing from ONT, plus increases in read depth/length/quality). </li>
            </ul>
          </li>
          <li>Note: in expanding to MANE Plus, MANE Select transcripts are expected to be stable (although we reserve the right to update them on the rare occasion that new data demonstrate the original determination is incorrect).</li>
        </ul>



    </div>

</div>
 </div>

{% endblock %}